WIP: Android support for gdc 4.8

mathstuf opened the pull request:
Currently a work in progress, but I'm filing a PR now to get information ahead of time. Subject to rebase/commit amending as necessary.

mathstuf committed:
build: Make phobos' configure depend on libiberty

The parallel build used in Android routinely failed without this.

mathstuf committed:
phobos: Look up another directory for libiberty

Android's NDK build scripts seem to have an extra level present here.

mathstuf commented:
I've gotten to the point where the only issue is that `__data_start` is not defined when linking a D executable. I'm not sure what is missing, but I can't get GDC to have it available (GCC is fine), so I'm guessing there's something missing behind the scenes that the linker scripts aren't being picked up (there's no argument differences for `collect2` or `ld` between the two except for the standard D library and include paths for GDC which GCC doesn't have).

mathstuf commented:
Seems the `__data_start` problem is related to the gold linker; using `ld.bfd` works just fine.

ibuclaw commented on libphobos/configure:
```diff
@@ -1,10 +1,10 @@
 #! /bin/sh
 # Guess values for system-dependent variables and create Makefiles.
-# Generated by GNU Autoconf 2.64 for libphobos version-unused.
+# Generated by GNU Autoconf 2.69 for libphobos version-unused.
```
I think this is a comment somewhere in the configure.ac file, but you need to install autoconf2.64 and rebuild the configure scripts using `autoreconf2.64`

mathstuf replied:
I didn't manually regenerate this; it was done by the NDK; the change I made to configure.ac could probably just be done manually to configure as well >.> .

ibuclaw commented on libphobos/libdruntime/core/runtime.d:
```diff
@@ -49,10 +49,14 @@ private
+    } else {
     version(GNU)
         import gcc.backtrace;
+    }
```
Ideally, you should instead be fixing  gcc.backtrace.

mathstuf replied:
The problem is actually that the backtrace implementations use stdout/stderr which are just /dev/null on Android, so there's not much to do here. I guess I could use the Android logging function, but that lives in `libandroid.so`.

jpf91 replied:
Is this also true if a program is executed via adb shell? IIRC it's possible to set the backtrace handler at runtime as well, so a higher level library could also open a window or something but that shouldn't be done in druntime. Using / linking libandroid.so might be acceptable though.

mathstuf replied:
It's not just that they're `/dev/null`, but `stdout` and `stderr` are not defined symbols. Anyways, this is for unit testing only, so not something I'd expect for final binaries.

ibuclaw commented on libphobos/libdruntime/core/sys/posix/pthread.d:
```diff
@@ -256,7 +256,36 @@ version( Posix )
+        }
+    }
+}
+else version( linux )
```
Someone's done a lot of this work upstream as Android x86 support in druntime (unfortunately, he didn't listen to me when I told him that his patches wouldn't work for GDC - both `linux` and `Android` are predefined versions on Android).  It may be worth looking at that so that you don't duplicate efforts.

ibuclaw commented on libphobos/libdruntime/core/sys/posix/pthread.d:
```diff
@@ -752,7 +781,10 @@ version( linux )
+    version( Android ) {
+    } else {
     int pthread_setschedprio(pthread_t, int);
+    }
```
Also, mumble mumble D-style coding conventions.

mathstuf replied:
Yeah, I wasn't sure what to do here. It seemed silly to duplicate the entire block to remove this, but indenting also looked weird.

jpf91 replied:
``` d
version( Android ) {}
else
    int pthread_setschedprio(pthread_t, int);
```

looks ok. not perfect, but acceptable imho ;-)

jpf91 replied:
But IIRC Walter usually wants these blocks to be duplicated.

ibuclaw commented on libphobos/libdruntime/core/sys/posix/signal_.c:
```diff
@@ -0,0 +1,44 @@
+
+int _sigaddset(sigset_t* set, int signum)
+{
+    return sigaddset(set, signum);
```
This is the lazy man's way to do it.  :o)

mathstuf replied:
What's the better way? They're not linkable since they're effectively macros :( .

jpf91 replied:
Porting the macros, if possible.

ibuclaw commented:
Look at these patches for druntime, as I'd not want you to do any more work than what is necessary.

https://github.com/D-Programming-Language/druntime/pull/734/files
https://github.com/D-Programming-Language/druntime/pull/738/files

mathstuf commented:
Yeah, it looks like he did a much more thorough sweep :) . Though that is also just x86. I'm looking for results sooner, so I'm just hitting the symbols I need…

jpf91 commented on libphobos/libdruntime/core/demangle.d:
```diff
@@ -440,7 +444,12 @@ private struct Demangle
-        val = strtold( tbuf.ptr, null );
+        version( Android ) {
+            // bionic does not have strtold.
+            val = atof( tbuf.ptr );
```
Is `strtod` also not available?

mathstuf replied:
Right. The man pages I have states it is equivalent to `strtod(str, NULL);` without the error checking.

jpf91 commented on libphobos/libdruntime/core/sys/posix/signal.d:
```diff
@@ -590,11 +590,35 @@ version( linux )
+        int sigismember(in sigset_t* set, int signum) {
+            return _sigismember(set, signum);
+        }
+    } else {
```
braces on separate lines please.

jpf91 commented:
Looks good so far, apart from some small nitpicks. However, almost all changes should be pushed to upstream druntime first.

mathstuf commented:
FWIW, https://github.com/mathstuf/druntime/tree/dev/android-gdc.

mathstuf commented:
And a PR filed. https://github.com/D-Programming-Language/druntime/pull/764. What's the standard process for updating druntime into GDC so I can test this better than "looks fine"?

mathstuf committed:
android: Backport Android support from druntime

mathstuf committed:
android: Special case File.seek

There's no fseek64 on Android, so just cast the value to an off_t. Since
Android doesn't support 64-bit addressing of files anyways, this
shouldn't be a real problem.

mathstuf committed:
android: Use GENERIC_IO, not GCC_IO

Android doesn't support the fget*_unlocked functions, so just emulate
them like on Windows.

mathstuf commented:
OK, so this is GDC with a druntime update to support the Android bits. There's an unresolved symbol `__emutls_v._tlsstart` when I start an app using this patchset, but it's much closer.

mathstuf committed:
android: Remove _tlsstart from core.thread

Android doesn't have it and since Windows is fine without, Android
should be as well.

mathstuf commented:
Whee. Now I'm up to crashes in the runtime (specifically `gc_qalloc`). Progress!

mathstuf committed:
android: add core.sys.elf and core.sys.link modules

jpf91 commented:
OK to close this?

mathstuf commented:
Yeah, 4.9 is the current target anyways.
